DIPs/10NNN-WGB.md

```diff
@@ -0,0 +1,173 @@
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+The current form of this is the automatic conversion of arguments to `lazy`
+parameters to delegates. The troubles with `lazy` are:
+
+1. it stands out as an oddity
+2. being an oddity means it is hard to reason about, especially with the
+proliferation of parameter attributes
+3. it is underdocumented
+4. special case code for it is sprinkled throughout the compiler
+5. it is rarely used, so likely has many undetected problems
+6. it has largely gone unrecognized that it works like a delegate
+
+`lazy` has found a home, however, in functions like this one in `std.path`:
+```
+pure @safe string absolutePath(string path, lazy string base = getcwd());
+```
+where it is undesirable to compute base argument unless it is actually needed.
+With this change, it could be rewritten as:
+
+```
+pure @safe string absolutePath(string path, string delegate() base = getcwd());
+```
+and called the same way.
+
+There's more than supplanting `lazy`. It makes delegates in general easier to write,
+and experience suggests that the easier they are to write, the more uses people
+find for them:
+
+```
+int delegate() dg = () { return 3; };
+```
+or:
+```
+int delegate() dg = () => 3;
+```
+become simply:
+```
+int delegate() dg = 3;
+```
+
+
+## Prior Work
+
+None known.
+
+
+## Description
+
+Allow the implicit conversion of an `Expression` of type `T` to a delegate lambda function that
+returns a type `T`. The body of the delegate lambda will be the `Expression` which
+will be returned from the lambda.
+
+I.e.:
+
+Given expression `E` of type `T`, it is converted to:
+
+```
+T delegate() { return E; }
+```
+
+Attribute inference is done on the function, just as it is for all lambda functions.
+
+The implicit conversion level will be "conversion".
+
+
+### Grammar Changes
+
+None.
+
+
+### Function Overloading
+
+```
+void biff(int);            // A
+void biff(int delegate()); // B
+
+void test(int i)
+{
+    biff(i);
+}
+```
+
+`i` is of type `int` which is an exact match for `A`. A match with `B` would be
+via a conversion to a delegate lambda, so `A` is better and is selected.
+
+```
+void biff(long);            // A
+void biff(long delegate()); // B
+
+void test(int i)
+{
+    biff(i);
+}
+```
+Both A and B are matched equally at the "conversion" level. Then, partial ordering is
+applied. A variable of type `long` can be converted to `long delegate()` lambda, but a
+`long delegate()` cannot be converted to `long`. Therefore, `A` is the better match.
+This makes sense as the intuitive result.
+
+
+### `null`
+
+```
+void biff(T delegate());
+
+void test()
+{
+    biff(null);
+}
+```
+`null` implicitly converts directly to any delegate, but is not of type `T`, so it will
+not be turned into a lambda. Again, this is what is expected and is consistent
+with existing code.
```

Comment:
It may be expected when passing a delegate, but it differs from expectations of a lazy parameter. This case breaks when switching to a delegate with implicit conversions:
```D
import std;

void foo(lazy int* ptr) {
    int* p = ptr; // with delegates, this segfaults
    writeln(p); // with lazy, this will print null
}

void main() {
    int i = 3;
    foo(null);
}
```

Reply:
You're right, but this DIP doesn't actually deprecate `lazy` yet so this is not a problem.

Reply:
Half of this DIP's rationale is based on 'supplanting `lazy`'. Let's say next month we're discussing a DIP deprecating `lazy` and this is brought up then. Are we going to retract this DIP? Do we keep around lazy anyway, undermining most of this DIP's rationale?